File: ruby/Problem04.rb
# Problem 4
# Find the largest palindrome made from the product of two 3-digit numbers.

num1 = (100..999).to_a.reverse
num2 = (100..999).to_a.reverse

products = []

num1.each do |num|
  num2.each_with_index do |item, index|
    products << num * item
  end
end

def is_palindrome(num)
  num.to_s == num.to_s.reverse
end

products.sort.reverse.find { |num| is_palindrome(num) }

File: python/Problem04.py
# Problem 4
# Find the largest palindrome made from the product of two 3-digit numbers.

def palindrome(x):
  return str(x) == str(x)[::-1]

a = 999
palindromes = []
while a >= 100:
  b = 999
  while b >= 100:
    if palindrome(a * b):
      palindromes.append(a * b)
    b -= 1
  a -= 1

max(palindromes)

max(
   a*b
   for a in range(100, 999)
   for b in range(100, 999)
   if palindrome(a*b)
)